File: Stack/496. Next Greater Element I.cpp
class Solution
{
public:
    vector<int> nextGreaterElement(vector<int> &nums1, vector<int> &nums2)
    {

        int n = nums2.size();
        vector<int> nge(n, -1);
        stack<int> st;

        for (int i = n - 1; i >= 0; i--)
        {

            while (!st.empty() && st.top() <= nums2[i])
                st.pop();

            if (!st.empty())
                nge[i] = st.top();

            st.push(nums2[i]);
        }

        int m = nums1.size();
        vector<int> ans(m);

        for (int i = 0; i < m; i++)
        {

            auto it = find(nums2.begin(), nums2.end(), nums1[i]);
            ans[i] = nge[it - nums2.begin()];
        }
        return ans;
    }
};

File: Stack/496. Next Greater Element I.java
class Solution {
    public int[] solve(int[] num2)
    {
        int next[]=new int[num2.length];
        Stack <Integer> st=new Stack<Integer>();
        for(int i=num2.length-1;i>=0;i--)
        {
            if(st.isEmpty())
                next[i]=-1;
            else
            {
                if(st.peek()>num2[i])
                {
                    next[i]=st.peek();
                }
                else
                {
                    while(st.size()>0 && st.peek()<=num2[i])
                    {
                        st.pop();
                    }
                    if(st.isEmpty())
                        next[i]=-1;
                    else
                    {
                        next[i]=st.peek();
                    }
                }
            }
            st.push(num2[i]);
        }
        return next;
    }
    public int[] nextGreaterElement(int[] num1, int[] num2) {
        int next[]=solve(num2);
        HashMap<Integer,Integer> mp= new HashMap<Integer,Integer>();
        for(int i=0;i<num2.length;i++)
        {
            mp.put(num2[i],next[i]);
        }
        int ans[]=new int[num1.length];
        for(int i=0;i<num1.length;i++)
        {
            if(mp.containsKey(num1[i]))
               ans[i]=mp.get(num1[i]);
        }
               return ans;
    }
}